compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Forwarding.hs

```diff
@@ -0,0 +1,174 @@
+instance LSP.HasTextDocument LSP.SignatureHelpParams LSP.TextDocumentIdentifier where
+  textDocument = signatureHelpParamsTextDocumentLens
+
+pullMonadThroughTuple :: Monad m => (a, m b) -> m (a, b)
+pullMonadThroughTuple (a, mb) = (a,) <$> mb
+
+-- Takes a natural transformation of responses and lifts it to forward the first error
+assumeSuccessCombiner
+  :: forall (m :: LSP.Method 'LSP.FromClient 'LSP.Request)
+  .  ([(FilePath, LSP.ResponseResult m)] -> LSP.ResponseResult m)
+  -> ResponseCombiner m
+assumeSuccessCombiner f res = f <$> mapM pullMonadThroughTuple res
+
+ignore :: Forwarding m
+ignore = ExplicitHandler $ \_ _ -> pure ()
+
+unsupported :: String -> Forwarding m
+unsupported name = ExplicitHandler $ \_ _ -> error $ "Attempted to call a method that is unsupported by the underlying IDEs: " <> name
+
+uriFilePathPrism :: Prism' LSP.Uri FilePath
+uriFilePathPrism = prism' LSP.filePathToUri LSP.uriToFilePath
+
+getMessageForwardingBehaviour
+  :: forall t (m :: LSP.Method 'LSP.FromClient t)
+  .  LSP.SMethod m
+  -> LSP.Message m
+  -> Forwarding m
+getMessageForwardingBehaviour meth params =
+  case meth of
+    LSP.SInitialize -> error "Forwarding of Initialize must be handled elsewhere."
+    LSP.SInitialized -> ignore
+    -- send to all then const reply
+    LSP.SShutdown -> ForwardRequest params $ AllRequest (assumeSuccessCombiner @m $ const LSP.Empty)
+    LSP.SExit -> ForwardNotification params AllNotification
+    LSP.SWorkspaceDidChangeWorkspaceFolders -> ForwardNotification params AllNotification
+    LSP.SWorkspaceDidChangeConfiguration -> ForwardNotification params AllNotification
+    LSP.SWorkspaceDidChangeWatchedFiles -> ForwardNotification params AllNotification
+    LSP.STextDocumentDidOpen -> ForwardNotification params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentDidChange -> ForwardNotification params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentWillSave -> ForwardNotification params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentWillSaveWaitUntil -> ForwardRequest params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentDidSave -> ForwardNotification params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentDidClose -> ForwardNotification params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentCompletion -> ForwardRequest params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentHover -> ForwardRequest params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentSignatureHelp -> ForwardRequest params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentDeclaration -> ForwardRequest params $ forwardingBehaviourFromParamsWithTextDocument params
+    LSP.STextDocumentDefinition -> error "Forwarding of STextDocumentDefinition must be handled elsewhere."
```

Comment:
TODO for future gardening work: make this type-level by having a type family that maps forwarding behaviour to Void for disallowed cases